tests/register/register.go

```diff
@@ -0,0 +1,258 @@
+	util.MustRun(t, "losetup", "-d", loopDevice)
+	test.RemoveAll(t, diskFile)
+}
+
+func (test Test) CreateDeviceMappers(t *testing.T, diskFile string) (devices []string) {
+	out := util.MustRun(t, "kpartx", "-avs", diskFile)
+	devices = util.RegexpSearchAll(t, "loop device", "map (?P<device>[\\w\\d]+)", out)
+
+	t.Logf("kpartx out: %s", string(out))
+
+	for i, d := range devices {
+		devices[i] = fmt.Sprintf("/dev/mapper/%s", d)
+	}
+	return
+}
+
+func (test Test) RemoveDeviceMappers(t *testing.T, diskFile string) {
+	util.MustRun(t, "kpartx", "-d", diskFile)
+}
+
+func (test Test) MountDeviceMapper(t *testing.T, device string) string {
+	dir, err := ioutil.TempDir("", "coreos-install-mount-point")
+	if err != nil {
+		t.Fatalf("couldn't create mount point directory: %v", err)
+	}
+
+	// use util.Run rather than util.MustRun so that we can send every
+	// partition through this method without special logic to filter out
+	// ones that don't have a filesystem
+	err = util.Run(t, "mount", device, dir, "-o", "ro")
+	if err != nil {
+		return ""
+	}
+
+	return dir
+}
+
+func (test Test) UnmountPath(t *testing.T, path string) {
+	util.MustRun(t, "umount", path)
+}
+
+func WhichCoreosInstall(t *testing.T) string {
+	out, err := exec.Command("which", "coreos-install").CombinedOutput()
+	if err != nil {
+		return ""
+	}
+
+	return filepath.Dir(string(out))
+}
+
+func (test Test) RunCoreOSInstall(t *testing.T, opts ...string) {
+	util.MustRun(t, "coreos-install", opts...)
+}
+
+func (test Test) ValidateIgnition(t *testing.T, mountPaths []string, config string) {
+	ignition_found := false
+	grub_found := false
+	for _, p := range mountPaths {
+		ignition_path := filepath.Join(p, "coreos-install.json")
+		if _, err := os.Stat(ignition_path); !os.IsNotExist(err) {
+			ignition_found = true
+			data, err := ioutil.ReadFile(ignition_path)
+			if err != nil {
+				t.Fatalf("couldn't read coreos-install.json: %v", err)
+			}
+
+			if string(data) != config {
+				t.Fatalf("coreos-install.json doesn't match: expected %s, received %s", config, data)
+			}
+
+		}
+
+		grub_path := filepath.Join(p, "grub.cfg")
+		if _, err := os.Stat(grub_path); !os.IsNotExist(err) {
+			grub_found = true
+			data, err := ioutil.ReadFile(grub_path)
+			if err != nil {
+				t.Fatalf("couldn't read grub.cfg: %v", err)
+			}
+
+			util.RegexpContains(t, "ignition grub.cfg info", "coreos.config.url=oem:///coreos-install.json", data)
+		}
+	}
+
+	if !ignition_found {
+		t.Fatalf("couldn't find coreos-install.json")
+	}
+
+	if !grub_found {
+		t.Fatalf("couldn't find grub.cfg")
+	}
+}
+
+func (test Test) RemoveAll(t *testing.T, path string) {
+	err := os.RemoveAll(path)
+	if err != nil {
+		t.Errorf("couldn't remove %s: %v", path, err)
+	}
+}
+
+func (test Test) WriteFile(t *testing.T, data string) string {
+	tmpFile, err := ioutil.TempFile("", "coreos-install-file")
+	if err != nil {
+		t.Fatalf("failed creating tmp file: %v", err)
+	}
```

Comment:
This function is used as a helper method when creating Ignition configs & Cloudconfigs, if I `defer` the `Close` call inside this method it will be deleted before `coreos-install` is ran.

As well, since this is inside the TempDir which has a defer'd RemoveAll should I worry about cleaning this file up outside of the outer RemoveAll?

Reply:
Ah, yeah.  Since you're inside a tmpdir anyway, and only need one cloud-config and one Ignition config per run, it should be okay to create a file with a fixed name (i.e. not using TempFile), close it before returning, and let the outer RemoveAll clean it up.  (Same with the disk image, for that matter.)